runners/core-java/src/main/java/org/apache/beam/runners/core/GroupAlsoByWindowViaWindowSetNewDoFn.java

```diff
@@ -0,0 +1,136 @@
+  private SideInputReader sideInputReader;
+  private DoFnRunners.OutputManager outputManager;
+  private TupleTag<KV<K, OutputT>> mainTag;
+
+  public GroupAlsoByWindowViaWindowSetNewDoFn(
+          WindowingStrategy<?, W> windowingStrategy,
+          StateInternalsFactory<K> stateInternalsFactory,
+          TimerInternals timerInternals,
+          SideInputReader sideInputReader,
+          SystemReduceFn<K, InputT, ?, OutputT, W> reduceFn,
+          DoFnRunners.OutputManager outputManager,
+          TupleTag<KV<K, OutputT>> mainTag) {
+    this.timerInternals = timerInternals;
+    this.sideInputReader = sideInputReader;
+    this.outputManager = outputManager;
+    this.mainTag = mainTag;
+    @SuppressWarnings("unchecked")
+    WindowingStrategy<Object, W> noWildcard = (WindowingStrategy<Object, W>) windowingStrategy;
+    this.windowingStrategy = noWildcard;
+    this.reduceFn = reduceFn;
+    this.stateInternalsFactory = stateInternalsFactory;
+  }
+
+  private OutputWindowedValue<KV<K, OutputT>> outputWindowedValue() {
+    return new OutputWindowedValue<KV<K, OutputT>>() {
+      @Override
+      public void outputWindowedValue(
+              KV<K, OutputT> output,
+              Instant timestamp,
+              Collection<? extends BoundedWindow> windows,
+              PaneInfo pane) {
+        outputManager.output(mainTag,
+                WindowedValue.of(output, timestamp, windows, pane));
+      }
+
+      @Override
+      public <SideOutputT> void sideOutputWindowedValue(
+              TupleTag<SideOutputT> tag,
+              SideOutputT output,
+              Instant timestamp,
+              Collection<? extends BoundedWindow> windows,
+              PaneInfo pane) {
+        outputManager.output(tag,
+                WindowedValue.of(output, timestamp, windows, pane));
+      }
+    };
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) throws Exception {
+    KeyedWorkItem<K, InputT> keyedWorkItem = c.element();
+
+    K key = keyedWorkItem.key();
+    StateInternals<K> stateInternals = stateInternalsFactory.stateInternalsForKey(key);
+
+    ReduceFnRunner<K, InputT, OutputT, W> reduceFnRunner =
+            new ReduceFnRunner<>(
+                    key,
+                    windowingStrategy,
+                    ExecutableTriggerStateMachine.create(
+                            TriggerStateMachines.stateMachineForTrigger(windowingStrategy.getTrigger())),
+                    stateInternals,
+                    timerInternals,
+                    outputWindowedValue(),
+                    sideInputReader,
+                    droppedDueToClosedWindow,
+                    reduceFn,
+                    c.getPipelineOptions());
+
+    reduceFnRunner.processElements(keyedWorkItem.elementsIterable());
+    reduceFnRunner.onTimers(keyedWorkItem.timersIterable());
+    reduceFnRunner.persist();
+  }
+
+  @Override
+  public OldDoFn<KeyedWorkItem<K, InputT>, KV<K, OutputT>> asDoFn() {
```

Comment:
Thanks for the review. Yes, I also wonder why DoFnRunners removed the createDefault, simpleRunner did not use lateDataDroppingRunner. If we discard too late data directly in the GroupAlsoByWindowViaWindowSetNewDoFn , will have any problems (such as with PushbackSideInputDoFnRunner)? Look forward to the subsequent changes.